### PR TITLE
Use torch.linalg.lstsq in InverseMelScale

### PR DESCRIPTION
**Pull request by nateanl**
To address https://github.com/pytorch/audio/issues/2643

### CODE REVIEW

File: test/torchaudio_unittest/transforms/transforms_test_impl.py [hunk 1]
```diff
@@ -15,7 +15,7 @@ def test_InverseMelScale(self):
         """Gauge the quality of InverseMelScale transform.
 
         As InverseMelScale is currently implemented with
-        random initialization + iterative optimization,
+        torch.linalg.lstsq optimization.
         it is not practically possible to assert the difference between
         the estimated spectrogram and the original spectrogram as a whole.
         Estimated spectrogram has very huge descrepency locally.
```
Comment (carolineechen) [inline test/torchaudio_unittest/transforms/transforms_test_impl.py:21]:
```suggestion
        Estimated spectrogram has very huge discrepancy locally.
```

**Review by carolineechen — COMMENTED** [review]
This looks like a pretty significant BC-breaking change that would break existing implementations using any of the removed parameters. Do you think it makes sense to maintain BC, by adding an option to choose which underlying implementation to use, with the old version being the default to avoid BC-breaking? Also normally for BC-breaking changes we will first issue a warning in a release before formally modifying it in the following release -- we could consider something like that if we are certain this approach is better/correct. Curious what the numerical difference looks like?

**Comment by nateanl** [comment]
> Curious what the numerical difference looks like?

The numerical difference is significant. So besides the API BC-breaking, it is also numerical BC-breaking.

**Comment by mthrok** [comment]
> This looks like a pretty significant BC-breaking change (or would this be considered bug-fix?) that would break existing implementations using any of the removed parameters. Do you think it makes sense to maintain BC, by adding an option to choose which underlying implementation to use, with the old version being the default to avoid BC-breaking? Also normally for BC-breaking changes we will first issue a warning in a release before formally modifying it in the following release -- we could consider something like that if we are certain this approach is better/correct. Curious what the numerical difference looks like?
> 
> cc @mthrok

Since the change is motivated by the improvement of numerical result, first we need to validate the improvement. Passing existing tests are not sufficient for this. Then we can think of a better way to handle interface breakage. If the new implementation should yield perfect inversion, then interface breakage is minor issue.

Firstly, per the following comment, the current implementation runs the optimizer, which generates blur result, and we cannot gauge the element-level differences.

https://github.com/pytorch/audio/blob/72404de901cdb47049d8f0383a4dc51a1b739317/test/torchaudio_unittest/transforms/transforms_test_impl.py#L17-L26

So the current test looks at the distribution of relative difference. We need to re-assess the numerical property of the new implementation and revise this to make sure that the change is improvement.

https://github.com/pytorch/audio/blob/72404de901cdb47049d8f0383a4dc51a1b739317/test/torchaudio_unittest/transforms/transforms_test_impl.py#L49-L53

Second issue is the possibility of the numerical instability. I do not know enough of the underlying math, but we need to test the new implementation on variety of configuration on different real/synthetic audio. We should be able to answer questions like, "is it reliably stable on common audios?", "if the input mel-spectrogram meets some conditions that make `lstsq` unstable, can we do something about it?"

**Comment by nateanl** [comment]
I doubt the numerical improvement would be marginal, stft -> mel is a lossy computation, hence both torchaudio and librosa use optimization method to estimate stft from mel spectrogram. The change in this PR mainly solves two issues: running speed and running in inference mode.

The [unit test in librosa](https://github.com/librosa/librosa/blob/f101492a5b6ee9ba09d12867b9718e58621003e3/tests/test_features.py#L814-L835) only compares the difference between `mel` and `mel_bank * stft`, where stft is the estimated one.
```python
def test_mel_to_stft(power, dtype, n_fft):
    srand()

    # Make a random mel spectrum, 4 frames
    mel_basis = librosa.filters.mel(sr=22050, n_fft=n_fft, n_mels=128, dtype=dtype)

    stft_orig = np.random.randn(n_fft // 2 + 1, 4) ** power
    mels = mel_basis.dot(stft_orig.astype(dtype))

    stft = librosa.feature.inverse.mel_to_stft(mels, power=power, n_fft=n_fft)

    # Check precision
    assert stft.dtype == dtype

    # Check for non-negative spectrum
    assert np.all(stft >= 0)

    # Check that the shape is good
    assert stft.shape[0] == 1 + n_fft // 2

    # Check that the approximation is good in RMSE terms
    assert np.sqrt(np.mean((mel_basis.dot(stft ** power) - mels) ** 2)) <= 5e-2
```

When comparing `stft` and `stft_orig` in librosa, the element-wise difference is also significant.

**Comment by nateanl** [comment]
Close it as `torch.linalg.lstsq` could return negative values in the output. We can still make it non-negative by using `torch.clamp`, but the result is not globally optimal.

**Pull request closed without merging**